File: src/Matlab/DataBinner.m
% Implementation is essentially identical to DataBinner.cpp/DataBinner.h
% TODO Learn to use c++ classes in Matlab to reduce code redundancy
classdef DataBinner < handle
    properties
        % For writing output
        fidOutput

        % Calculation variables
        windowLength {mustBeNumeric}
        binLength {mustBeNumeric}
        nBins {mustBeNumeric}
        nextBinTime {mustBeNumeric}
        currentIndex {mustBeNumeric}
        
        % Cells of maps
        bins
    end

    methods
        function obj = DataBinner(outputFileName, windowLength,binLength)
            obj.fidOutput = fopen(outputFileName, 'w');

            obj.windowLength = windowLength;
            obj.binLength = binLength;
            obj.nBins = windowLength / binLength;
            obj.nextBinTime = binLength;
            obj.currentIndex = 1;

            % Initialize bins with empty maps
            create_containers = @(n)arrayfun(@(x)containers.Map('KeyType', 'uint32', 'ValueType', 'double'), 1:n, 'UniformOutput', false);
            obj.bins = create_containers(obj.nBins);
        end


        function insert(obj, times, channels)
            % Iterate over the spikes
            for i = 1 : size(times, 2)
                % If data belongs to next bin
                if times(i) > obj.nextBinTime
                    % Set index to oldest bin
                    obj.currentIndex = mod((obj.currentIndex), obj.nBins) + 1;
                    
                    % Clear oldest bin for new data
                    remove(obj.bins{obj.currentIndex}, keys(obj.bins{obj.currentIndex}));

                    % Update next subbin time
                    obj.nextBinTime = obj.nextBinTime + obj.binLength;
                end

                % Increment spike count of the channel by 1
                incrementMap(obj.bins{obj.currentIndex}, channels(i), 1);
            end
        end

        function window = getDataWindow(obj)
            % Set window as empty map
            window = containers.Map('KeyType', 'uint32', 'ValueType', 'double');
            
            % Iterate over bins
            for i = 1 : obj.nBins
                % Iterate over bin's channel counts
                for channel = keys(obj.bins{i})
                    count = obj.bins{i}(channel{1});
                    incrementMap(window, channel{1}, count);
                end
            end
        end

        function writeDecoderInput(obj, label, window)
            fprintf(obj.fidOutput, '+%d', label);
            for channel = keys(window)
                count = window(channel{1});
                fprintf(obj.fidOutput, ' %d:%f', channel{1} + 1, count); % Channel +1 because libSVM is 1-indexed
            end
            fprintf(obj.fidOutput, '\n');
        end

        function readInSpikes(obj, spikeTimes, spikeTemplates, eventFileName)
            fid = fopen(eventFileName);
            events = fscanf(fid, '%d %d', [2, inf]);
            fclose(fid);

            eventTimes = events(1, :);
            eventLabels = events(2, :);

            times = [];
            channels = [];
            eventIdx = 1;
            
            for i = 1:size(spikeTimes, 1)
                % If a window's length of time has passed since the time
                % vent, record dataWindow and extract a new label
                if eventTimes(eventIdx) + obj.windowLength < spikeTimes(i)
                    % Insert Times and channels vectors into the binner
                    insert(obj, times, channels);
                    times = [];
                    channels = [];

                    window = getDataWindow(obj);

                    writeDecoderInput(obj, eventLabels(eventIdx), window);
                    
                    % Set eventIdx for the next event and its label
                    eventIdx = eventIdx + 1;

                    if eventIdx > size(eventTimes, 2)
                        break
                    end
                end

                % Insert time and channel
                times = [times spikeTimes(i)];
                channels = [channels spikeTemplates(i)];
            end

            fclose(obj.fidOutput);
        end
    end
end

% Helper function
function incrementMap(map, key, add)
    if isKey(map, key) % If key exists in map
        map(key) = map(key) + add;
    else % If key doesn't exist in map
        map(key) = add;
    end
end

File: src/Decoder/dataBinner.cpp
#include "dataBinner.h"
#include <thread>
#include "Decoder.h"
#include "../Networking/NetworkHelpers.h"
#include "../Helpers/TimeHelpers.h"

DataBinner::DataBinner(int windowLength, int binLength, int windowOffset)
	: windowLength(windowLength)
	, binLength(binLength)
	, windowOffset(windowOffset)
	, nBins(windowLength / binLength + 1)
	, nextBinTime(binLength)
	, currentIndex(0)
{
	// Initialize bins with empty maps
	for (int i = 0; i < nBins; i++) {
		std::map<long, double> bin;
		bins.push_back(bin);
	}
};

DataBinner::~DataBinner() {};

void DataBinner::insert(std::vector<long> times, std::vector<long> channels) {
	std::lock_guard lk(guard);

	// Iterate over the spikes
	for (int i = 0; i < times.size(); i++) {
		// Update bins to new data, no assumptions on gaps or anything, as long as data comes in monotonically
		while (times[i] > nextBinTime) {
			// Set index to oldest bin
			currentIndex = (currentIndex + 1) % nBins;

			// Clear oldest bin for new data
			bins[currentIndex].clear();
			
			// Update next subbin time
			nextBinTime += binLength;
		}

		// Increment spike count of the channel by 1 
		// If the spike time received is too old so as to not make sense in circular buffer, ignore
		if ((nextBinTime - times[i]) / binLength < nBins) {
			// ex: d = (nextBinTime - times[i]) / binLength = how many indices behind currentIndex times[i] should be in
			//    Thus, we should place times[i] into (currentIndex - d) mod nBins
			//    However, in C++ -7 % 2 == -2, and so we adjust for this by doing (nBins + (x % nBins)) % nBins
			//    instead of just x % nBins
			bins[(nBins + (currentIndex - (nextBinTime - times[i]) / binLength) % nBins) % nBins][channels[i]]++;
		}
		else {
			std::cout << "Spike with spike time = " << times[i] << " discarded, as it is too old." << std::endl;
		}
	}
}

void DataBinner::updateTime(long time) 
{
	std::lock_guard lk(guard);
	while (time > nextBinTime) {
		currentIndex = (currentIndex + 1) % nBins;
		bins[currentIndex].clear();
		nextBinTime += binLength;
	}
}

// Return bin map that is the accumulation of all the subbin maps
std::map<long, double> DataBinner::getDataWindow() {
	std::lock_guard lk(guard);

	// Set window as copy of first bin 
	std::map<long, double> window = bins[0];

	// Iterate over bins
	for (int i = 1; i < nBins; i++) {
		// Iterate over bin's channel counts
		for (auto const& [channel, count] : bins[i]) {
			window[channel] += count;
		}
	}

	return window;
}


void DataBinner::readInSpikes(const char* csvFilePath, const char* eventFilePath, const char* outputFilePath, FileWriter *m_fwOut) {
	std::vector<long> times, channels;
	std::cout << "Binner reading in from " << csvFilePath << " and " << eventFilePath << std::endl;
	FILE *fpIn;
	FILE *fpLabel;
	FILE *fpOut;

	if ((fpIn = fopen(csvFilePath, "r")) == NULL)
		fprintf(stderr, "Can't open input file %s\n", csvFilePath);
	if ((fpLabel = fopen(eventFilePath, "r")) == NULL)
		fprintf(stderr, "Can't open input file %s\n", eventFilePath);
	fpOut = fopen(outputFilePath, "w");

	// Extract label and next label
	int eventLabel;
	int eventTime;
	fscanf(fpLabel, "%d %d", &eventTime, &eventLabel);

	long time, channel;
	float amplitude; // Thrown away

	while (fscanf(fpIn, "%20ld,%4ld,%f", &time, &channel, &amplitude) == 3)
	{
		// If a window's length of time has passed since the time event, record dataWindow and extract a new label
		if (eventTime + windowOffset < time) {
			if (times.size() > 0) {
				// Insert times and channels vectors into the data binner
				insert(times, channels);
				updateTime(eventTime);
				times.clear();
				channels.clear();

				std::map<long, double> window = getDataWindow();
				m_fwOut->WriteDecoderInput(eventLabel, window);
			}
			else {	
				std::cout << "Skipping event " << eventTime << std::endl;
			}
			// Extract the time of the next event and its label
			if (fscanf(fpLabel, "%d %d", &eventTime, &eventLabel) != 2)
				break;
		}

		// Insert time and channel
		times.push_back(time);
		channels.push_back(channel);
	}
	fclose(fpIn);
	fclose(fpLabel);
	fclose(fpOut);

	// Reset databinner for real time sorting
	for (int i = 0; i < nBins; i++) {
		bins[i].clear();
	}
	currentIndex = 0;
	nextBinTime = binLength;
}

/*-
 * Copyright (c) 1990, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */
char* DataBinner::strsep(char **stringp, const char *delim) {
	char *s;
	const char *spanp;
	int c, sc;
	char *tok;

	if ((s = *stringp) == NULL)
		return (NULL);
	for (tok = s;;) {
		c = *s++;
		spanp = delim;
		do {
			if ((sc = *spanp++) == c) {
				if (c == 0)
					s = NULL;
				else
					s[-1] = 0;
				*stringp = s;
				return (tok);
			}
		} while (sc != 0);
	}
}